src/Microsoft.IdentityModel.Clients.ActiveDirectory/Internal/Flows/AcquireTokenHandlerBase.cs

```diff
@@ -205,10 +223,25 @@ private void StoreResultExToCache(ref bool notifiedBeforeAccessCache)
                     this.NotifyBeforeAccessCache();
                     notifiedBeforeAccessCache = true;
                 }
-
-                this.tokenCache.StoreToCache(ResultEx, this.Authenticator.Authority, this.Resource,
+                var metadata = await InstanceDiscovery.GetMetadataEntry(
+                    GetHost(this.Authenticator.Authority), this.Authenticator.ValidateAuthority, this.CallState).ConfigureAwait(false);
+                this.tokenCache.StoreToCache(
+                    ResultEx, ReplaceHost(this.Authenticator.Authority, metadata.PreferredCache), this.Resource,
                     this.ClientKey.ClientId, this.TokenSubjectType, this.CallState);
             }
+            return notifiedBeforeAccessCache;
+        }
+
+        private string GetHost(string uri)
+        {
+            // Note: host is supposed to be case insensitive, but we would like to preserve its case to match a previously cached token
+            return uri.Split('/')[2];
```

Comment:
hm that seems a bit dangerous... don't we want to do some validation that this uri is actually a uri, and is valid?

Reply:
Solution 1: Well, if you prefer, we can add a `new Uri(uri)` before this line, so that all the invalid input that fails [the default checks in Uri() constructor](https://msdn.microsoft.com/en-us/library/z6c2z492(v=vs.110).aspx#Exceptions) would end up with an (different) exception.

Solution 2: the reason that I did not go with the elagent `new Uri(uri).Host` implementation is that, this [pre-existing line in ANOTHER test case](https://github.com/AzureAD/azure-activedirectory-library-for-dotnet/blob/Release-3.17.0/tests/Test.ADAL.Common/TokenCacheTests.cs#L177) somehow uses mixed case as input. So after the implicit normalization of `new Uri(uri).Host`, it gives us a lower case host name which won't match the previous record in the cache. Note that, in practice, host would and probably should always be in lower case, so this problem should not exist. If you don't mind, I can go ahead and adjust that aforementioned test case, and then we can use `new Uri(uri).Host` here.

What do you think?